strawberryfields/backends/tfbackend/circuit.py

```diff
@@ -360,121 +343,110 @@ def prepare_multimode(self, state, modes, input_state_is_pure=False):
         """
-        with self._graph.as_default():
-            k = tf.cast(kappa, ops.def_type)
-            k = self._maybe_batch(k)
-            new_state = ops.cross_kerr_interaction(k, mode1, mode2, self._state, self._cutoff_dim, self._state_is_pure, self._batched)
-            self._update_state(new_state)
+        k = tf.cast(kappa, ops.def_type)
+        k = self._maybe_batch(k)
+        new_state = ops.cross_kerr_interaction(k, mode1, mode2, self._state, self._cutoff_dim, self._state_is_pure, self._batched)
+        self._update_state(new_state)
 
     def cubic_phase(self, gamma, mode):
         """
         Apply the cubic phase operator to the specified mode.
         """
-        with self._graph.as_default():
-            g = tf.cast(gamma, ops.def_type)
-            g = self._maybe_batch(g)
-            new_state = ops.cubic_phase(g, mode, self._state, self._cutoff_dim, self._hbar, self._state_is_pure, self._batched)
-            self._update_state(new_state)
+        g = tf.cast(gamma, ops.def_type)
+        g = self._maybe_batch(g)
+        new_state = ops.cubic_phase(g, mode, self._state, self._cutoff_dim, self._hbar, self._state_is_pure, self._batched)
+        self._update_state(new_state)
 
     def loss(self, T, mode):
         """
         Apply a loss channel  to the specified mode.
         """
-        with self._graph.as_default():
-            # in_modes = self._state
-            T = tf.cast(T, ops.def_type)
-            T = self._maybe_batch(T)
-            new_state = ops.loss_channel(T, mode, self._state, self._cutoff_dim, self._state_is_pure, self._batched)
-            self._update_state(new_state)
-            self._state_is_pure = False # loss output always in mixed state representation
+        # in_modes = self._state
+        T = tf.cast(T, ops.def_type)
+        T = self._maybe_batch(T)
+        new_state = ops.loss_channel(T, mode, self._state, self._cutoff_dim, self._state_is_pure, self._batched)
+        self._update_state(new_state)
+        self._state_is_pure = False # loss output always in mixed state representation
 
     def vacuum_element(self):
         """Compute the fidelity with the (multi-mode) vacuum state."""
-        with self._graph.as_default():
-            if self._batched:
-                vac_component = tf.reshape(self._state, [self._batch_size, -1])[:, 0]
-            else:
-                vac_component = tf.reshape(self._state, [-1])[0]
+        if self._batched:
+            vac_component = tf.reshape(self._state, [self._batch_size, -1])[:, 0]
+        else:
+            vac_component = tf.reshape(self._state, [-1])[0]
 
-            if self._state_is_pure:
-                return tf.abs(vac_component) ** 2
-            return vac_component
+        if self._state_is_pure:
+            return tf.abs(vac_component) ** 2
+        return vac_component
 
     def measure_fock(self, modes, select=None, **kwargs):
```

Comment:
Just a hunch, but I have a feeling the measurement ops will give us the biggest headaches when we push for tf.autograph compatibility

Reply:
How so?
In any case, the best way to be safe I think is to use stock TF functions wherever possible.

Reply:
Just a future thought, but if we go down the python `@tf.custom_gradient` route with the gates, autograph won't be supported. I think you need to implement the TF API directly in C++ to get autograph support.